lib/internal/modules/esm/module_map.js

```diff
@@ -1,17 +1,119 @@
 'use strict';
 
-const { kImplicitAssertType } = require('internal/modules/esm/assert');
 const {
+  ArrayPrototypeJoin,
+  ArrayPrototypeMap,
+  ArrayPrototypeSort,
+  JSONStringify,
+  ObjectDefineProperty,
+  ObjectKeys,
   SafeMap,
 } = primordials;
+const { kImplicitAssertType } = require('internal/modules/esm/assert');
+const { setOwnProperty } = require('internal/util');
 let debug = require('internal/util/debuglog').debuglog('esm', (fn) => {
   debug = fn;
 });
 const { ERR_INVALID_ARG_TYPE } = require('internal/errors').codes;
 const { validateString } = require('internal/validators');
 
-// Tracks the state of the loader-level module cache
-class ModuleMap extends SafeMap {
+/**
+ * Cache the results of the `resolve` step of the module resolution and loading process.
+ * Future resolutions of the same input (specifier, parent URL and import assertions)
+ * must return the same result if the first attempt was successful, per
+ * https://tc39.es/ecma262/#sec-HostLoadImportedModule.
+ */
+class ResolveCache extends SafeMap {
+  constructor(i) { super(i); } // eslint-disable-line no-useless-constructor
+
+  /**
+   * Generates the internal serialized cache key and returns it along the actual cache object.
+   *
+   * It is exposed to allow more efficient read and overwrite a cache entry.
+   * @param {string} specifier
+   * @param {string} parentURL
+   * @param {Record<string,string>} importAssertions
+   * @returns {{
+   *   serializedModuleRequest: string,
+   *   internalCache: Record<string, Promise<import('./loader').ModuleExports> | import('./loader').ModuleExports>,
```

Comment:
It might be worth entirely separating the serialization process from the lookup process?

Something like:

```js
const requestKey = resolveCache.serializeKey(specifier, attributes);
const absoluteResolvedURL = await resolve(specifier, parentURL, attributes);
const resolvedKey = resolveCache.serializeKey(absoluteResolvedURL, attributes);
resolveCache.set(requestKey, parentKey, resolvedKey);
```

Ideally we should also be using `parentKey` not `parentURL` I think. And then note that you probably want to avoid repeat serialization so you want to be maintaining these keys on the module jobs themselves for easy lookup.

Does that make sense? Let me know if I can clarify any of the above.

Reply:
> Ideally we should also be using `parentKey` not `parentURL` I think

That's a good remark, I didn't consider this. I don't think we have access to the `parentKey` when doing dynamic imports, do we? 🤔 Using `parentURL` doesn't go against the spec IIUC, it does add a limitation that's not in the spec (but it likely won't matter that much in practice, I'd expect most use case would be more than fine if modules that have the same URL use the same resolve cache), so I'm tempted to say `parentURL` is good enough. wdyt?

> It might be worth entirely separating the serialization process from the lookup process?

That's probably a good idea. The current implementation is probably not going to stay as Jacob is trying to refactor all that is his PR, so I'm hesitant on spending too much time crafting an implementation that we want to get rid of anyway. What I'm mostly interested in is to get the tests to land, so we can refactor with more confidence.

Reply:
What is `parentKey` exactly? A serialized `parentURL` plus assertions/attributes?